### PR TITLE
[stdlib] Improved documentation for popLast() & removeLast() in BidirectionalCollection

### PR DESCRIPTION
**Pull request by louisdh**
Added detailed documentation to `popLast()` regarding its safety compared to `removeLast()`,
as well as noting the performance cost `popLast()` has.

### CODE REVIEW

**Comment by hooman** [comment]
I think a similar note is needed for `removeLast()` as well.

**Comment by LouisDhauwe** [comment]
@hooman Good point, documentation has now been updated for both `popLast()` and `removeLast()`.

**Review by dabrahams — COMMENTED** [review]
These changes are good, but we shouldn't be claiming popLast is safer. It's only preferable because it's guaranteed to be O(1) if it's available.

**Comment by LouisDhauwe** [comment]
@dabrahams Ok, I removed the safety claim.

**Comment by dabrahams** [comment]
I'm sorry, but I'm not comfortable with the recommendation either.  It seems to me that either `removeLast` should be considered no less preferable than `removeRange`, or if we really believe its usefulness is so marginal as to be a corner case, we should just tell users to `remove(at: index(before: endIndex))`.

**Comment by moiseev** [comment]
@dabrahams Safety claim is perhaps my incorrect wording. It is safer in a sense that the type gives you all the info about the behavior of the function, whereas with `removeLast` it will trap in runtime. It is not about memory safety, it's about convenience and predictable behavior.

File: stdlib/public/core/BidirectionalCollection.swift [hunk 1]
```diff
@@ -204,6 +204,11 @@ extension BidirectionalCollection where SubSequence == BidirectionalSlice<Self>
 
 extension BidirectionalCollection where SubSequence == Self {
   /// Removes and returns the last element of the collection.
+  /// 
+  /// `popLast()` should generally be preferred over `removeLast()`.
+  /// 
+  /// You should only opt for `removeLast()` if you are certain the collection 
+  /// is not empty and performance is an issue.
```
Comment (natecook1000) [inline stdlib/public/core/BidirectionalCollection.swift]: Could we rewrite this as:

> Use `popLast()` to remove the last element of a collection that might be empty. The `removeLast()` method must be used only on a nonempty collection.

**Review by natecook1000 — CHANGES_REQUESTED** [review]

**Comment by LouisDhauwe** [comment]
@natecook1000 Ok, I've updated according to your requested changes.

**Comment by moiseev** [comment]
Sorry for the back and forth on this, the problem turned out to be deeper than it originally looked =)

The problem with the comment I originally suggested is: both these methods are not necessarily available in all contexts. You can see that `removeLast` is defined for bidirectional collections and range-replaceable collections, but `popLast` is only available in bidirectional.

Interestingly, `removeLast` uses a hook in the protocol (see `_customRemoveLast`) that returns the last element if it existed, but it also requires the collection to be non-empty.

TL;DR: I think the right thing to do *now* is simply cross-reference the two methods in the documentation. Pretty much the way you've done it at the time of this writing.

This case clearly identifies the flaw in the design of these methods, solving which might become a part of a follow-up to [this proposal](https://github.com/apple/swift-evolution/blob/master/proposals/0132-sequence-end-ops.md) one day.

Thanks for working on this, @LouisDhauwe!

**Comment by moiseev** [comment]
This is sort of weird for a comment only change, but needs to be done.
@swift-ci Please smoke test

**Review by natecook1000 — APPROVED** [review]

**Comment by moiseev** [comment]
@swift-ci Please smoke test and merge

**Comment by moiseev** [comment]
@swift-ci Please smoke test and merge
